File: tests/checker.test.ts
import { describe, expect, it } from "vitest";
import {
  check,
  getExprType,
  getFieldInfo,
  getFunctionInfo,
  getLetType,
  getStructInfo
} from "../src/typeck/checker.js";
import type { BinaryExpression, LetStatement } from "../src/parser/ast.js";
import { SourceFile } from "../src/source/source-file.js";

function checkSource(text: string) {
  return check(new SourceFile("test.ik", text));
}

function messagesOf(text: string): string[] {
  return checkSource(text).diagnostics.map((diagnostic) => diagnostic.message);
}

interface NegativeCheckerCase {
  name: string;
  source: string;
  keywords: string[];
}

const negativeCheckerCases: NegativeCheckerCase[] = [
  {
    name: "unknown variable",
    source: `
      export fn bad(a: i32) -> i32 {
        return x;
      }
    `,
    keywords: ["unknown", "variable", "x"]
  },
  {
    name: "unknown function",
    source: `
      export fn bad(a: i32) -> i32 {
        return missing(a);
      }
    `,
    keywords: ["unknown", "function", "missing"]
  },
  {
    name: "unknown struct type",
    source: `
      export fn bad(x: ptr<Missing>) -> i32 {
        return 0;
      }
    `,
    keywords: ["unknown", "type", "Missing"]
  },
  {
    name: "unknown struct field",
    source: `
      struct Item {
        price: i64;
      }

      export fn bad(item: ptr<Item>) -> i64 {
        return item[0].qty;
      }
    `,
    keywords: ["field", "qty"]
  },
  {
    name: "duplicate struct name",
    source: `
      struct Item { price: i64; }
      struct Item { qty: i64; }
    `,
    keywords: ["duplicate", "struct", "Item"]
  },
  {
    name: "duplicate function name",
    source: `
      export fn calc() -> i32 { return 1; }
      export fn calc() -> i32 { return 2; }
    `,
    keywords: ["duplicate", "function", "calc"]
  },
  {
    name: "duplicate variable name",
    source: `
      export fn bad() -> i32 {
        let x: i32 = 1;
        let x: i32 = 2;
        return x;
      }
    `,
    keywords: ["duplicate", "variable", "x"]
  },
  {
    name: "wrong return type",
    source: `
      export fn bad() -> i32 {
        return true;
      }
    `,
    keywords: ["return", "expected", "i32", "bool"]
  },
  {
    name: "if condition is not bool",
    source: `
      export fn bad(a: i32) -> i32 {
        if a {
          return 1;
        }
        return 0;
      }
    `,
    keywords: ["if", "condition", "bool", "i32"]
  },
  {
    name: "while condition is not bool",
    source: `
      export fn bad(a: i32) -> i32 {
        while a {
          return 1;
        }
        return 0;
      }
    `,
    keywords: ["while", "condition", "bool", "i32"]
  },
  {
    name: "i32 plus i64 is not allowed",
    source: `
      export fn bad(a: i32, b: i64) -> i32 {
        return a + b;
      }
    `,
    keywords: ["arithmetic", "+", "same", "type"]
  },
  {
    name: "bool plus i32 is not allowed",
    source: `
      export fn bad(a: bool, b: i32) -> i32 {
        return a + b;
      }
    `,
    keywords: ["arithmetic", "+", "integer"]
  },
  {
    name: "wrong function argument count",
    source: `
      export fn add(a: i32, b: i32) -> i32 {
        return a + b;
      }

      export fn bad() -> i32 {
        return add(1);
      }
    `,
    keywords: ["function", "add", "expects", "2", "got", "1"]
  },
  {
    name: "wrong function argument type",
    source: `
      export fn add(a: i32, b: i32) -> i32 {
        return a + b;
      }

      export fn bad() -> i32 {
        return add(1, true);
      }
    `,
    keywords: ["argument", "2", "add", "expects", "i32", "bool"]
  },
  {
    name: "pointer index type is not i32 or u32",
    source: `
      struct Item {
        price: i64;
      }

      export fn bad(items: ptr<Item>, idx: bool) -> i64 {
        return items[idx].price;
      }
    `,
    keywords: ["index", "i32", "u32", "bool"]
  },
  {
    name: "assignment target is an integer literal",
    source: `
      export fn bad() -> i32 {
        1 = 2;
        return 0;
      }
    `,
    keywords: ["assignment", "target"]
  },
  {
    name: "assignment target is a function call",
    source: `
      export fn foo() -> i32 {
        return 1;
      }

      export fn bad() -> i32 {
        foo() = 2;
        return 0;
      }
    `,
    keywords: ["assignment", "target"]
  },
  {
    name: "assignment type mismatch",
    source: `
      export fn bad() -> i32 {
        let x: i32 = 1;
        x = true;
        return x;
      }
    `,
    keywords: ["assign", "bool", "i32"]
  },
  {
    name: "missing return",
    source: `
      export fn bad(a: i32) -> i32 {
        let x: i32 = a;
      }
    `,
    keywords: ["missing", "return"]
  },
  {
    name: "duplicate struct field",
    source: `
      struct Item {
        price: i64;
        price: i64;
      }
    `,
    keywords: ["duplicate", "field", "price"]
  }
];

function expectDiagnosticWithKeywords(source: string, keywords: string[]) {
  const diagnostics = checkSource(source).diagnostics;
  const matchingDiagnostic = diagnostics.find((diagnostic) => {
    const message = diagnostic.message.toLowerCase();
    return keywords.every((keyword) => message.includes(keyword.toLowerCase()));
  });

  expect(diagnostics.length, `expected at least one diagnostic for keywords: ${keywords.join(", ")}`).toBeGreaterThan(0);
  expect(
    matchingDiagnostic,
    `expected diagnostic containing keywords ${keywords.join(", ")} in messages:\n${diagnostics.map((diagnostic) => diagnostic.message).join("\n")}`
  ).toBeDefined();
  expect(matchingDiagnostic?.line).toEqual(expect.any(Number));
  expect(matchingDiagnostic?.column).toEqual(expect.any(Number));
  expect(matchingDiagnostic?.line).toBeGreaterThan(0);
  expect(matchingDiagnostic?.column).toBeGreaterThan(0);
}

describe("checker", () => {
  it("accepts a valid pricing-style program and records expression types", () => {
    const result = checkSource(`
      struct Item {
        price: i64;
        qty: i64;
        tax_rate_ppm: i64;
      }

      export fn tax(base: i64, ppm: i64) -> i64 {
        return base * ppm / 1000000;
      }

      export fn calc(items: ptr<Item>, len: i32, out: ptr<i64>) -> i32 {
        let i: i32 = 0;
        while i < len {
          let subtotal: i64 = items[i].price * items[i].qty;
          if subtotal > 0 {
            out[i] = subtotal + tax(subtotal, items[i].tax_rate_ppm);
          } else {
            out[i] = 0;
          }
          i = i + 1;
        }
        return 0;
      }
    `);

    expect(result.diagnostics).toEqual([]);
    expect(result.typedAst.program.kind).toBe("Program");
    expect([...result.typedAst.expressionTypes.values()].some((type) => type.kind === "primitive" && type.name === "bool")).toBe(true);
    expect([...result.typedAst.expressionTypes.values()].some((type) => type.kind === "pointer")).toBe(true);
  });

  it("exposes a stable checked program contract for MIR lowering", () => {
    const result = checkSource(`
      struct Item {
        price: i64;
        qty: i64;
      }

      export fn add(a: i64, b: i64) -> i64 {
        return a + b;
      }

      export fn calc(item: ptr<Item>) -> i64 {
        let subtotal: i64 = item[0].price + add(1, 2);
        return subtotal;
      }
    `);

    expect(result.diagnostics).toEqual([]);

    const checkedProgram = result.checkedProgram;
    const addInfo = getFunctionInfo(checkedProgram, "add");
    const calcInfo = getFunctionInfo(checkedProgram, "calc");
    const itemInfo = getStructInfo(checkedProgram, "Item");
    const priceInfo = getFieldInfo(checkedProgram, "Item", "price");

    expect(addInfo?.exported).toBe(true);
    expect(addInfo?.returnType).toEqual({ kind: "primitive", name: "i64" });
    expect(addInfo?.params.map((param) => [param.name, param.type])).toEqual([
      ["a", { kind: "primitive", name: "i64" }],
      ["b", { kind: "primitive", name: "i64" }]
    ]);
    expect(calcInfo?.params[0]?.type).toEqual({
      kind: "pointer",
      elementType: { kind: "struct", name: "Item" }
    });
    expect(itemInfo?.fields.map((field) => field.name)).toEqual(["price", "qty"]);
    expect(priceInfo?.type).toEqual({ kind: "primitive", name: "i64" });

    const calcDeclaration = calcInfo?.declaration;
    expect(calcDeclaration).toBeDefined();
    const letStatement = calcDeclaration!.body.statements[0] as LetStatement;
    const initializer = letStatement.initializer as BinaryExpression;

    expect(getLetType(checkedProgram, letStatement)).toEqual({ kind: "primitive", name: "i64" });
    expect(getExprType(checkedProgram, initializer)).toEqual({ kind: "primitive", name: "i64" });
    expect(getExprType(checkedProgram, initializer.left)).toEqual({ kind: "primitive", name: "i64" });
    expect(getExprType(checkedProgram, initializer.right)).toEqual({ kind: "primitive", name: "i64" });
  });

  it("reports duplicate declarations, variables, and unknown names", () => {
    expect(messagesOf(`
      struct Item { price: i64; }
      struct Item { qty: i64; }

      export fn f(a: i32, a: i32) -> i32 {
        let x: Missing = 0;
        let x: i32 = y;
        return x;
      }

      export fn f() -> i32 {
        return 0;
      }
    `)).toEqual(
      expect.arrayContaining([
        "Duplicate struct 'Item'.",
        "Duplicate function 'f'.",
        "Duplicate variable 'a'.",
        "Unknown type 'Missing'.",
        "Duplicate variable 'x'.",
        "Unknown variable 'y'."
      ])
    );
  });

  it("reports field, call, condition, return, assignment, arithmetic, and index errors", () => {
    expect(messagesOf(`
      struct Item { price: i64; }

      export fn takes_i64(value: i64) -> i64 {
        return value;
      }

      export fn bad(items: ptr<Item>, flag: bool, idx64: i64) -> i32 {
        let value: i64 = items[idx64].missing;
        let mixed: i64 = value + flag;
        if value {
          value = flag;
        } else {
          value = takes_i64(flag, value);
        }
        while 1 {
          value = value + 1;
        }
        return value == 0;
      }
    `)).toEqual(
      expect.arrayContaining([
        "Index expression requires i32 or u32 index, got i64.",
        "Struct 'Item' has no field 'missing'.",
        "Arithmetic operator '+' requires integer operands of the same type.",
        "If condition must be bool, got i64.",
        "Cannot assign bool to i64.",
        "Function 'takes_i64' expects 1 argument but got 2.",
        "Argument 1 of function 'takes_i64' expects i64 but got bool.",
        "While condition must be bool, got i32.",
        "Return type mismatch: expected i32 but got bool."
      ])
    );
  });

  it("reports diagnostics with line and column", () => {
    const result = checkSource(`
      export fn bad() -> i32 {
        return missing;
      }
    `);

    expect(result.diagnostics).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          code: "IK2001",
          severity: "error",
          message: "Unknown variable 'missing'.",
          fileName: "test.ik",
          line: 3,
          column: 16
        })
      ])
    );
    expect(result.diagnostics[0]?.span).toMatchObject({
      start: { line: 3, column: 16 }
    });
  });

  describe("negative cases", () => {
    it.each(negativeCheckerCases)("$name", ({ source, keywords }) => {
      expectDiagnosticWithKeywords(source, keywords);
    });
  });
});

File: tests/checker_test.rs
use calckernel::{
    CalcKernelType, DiagnosticCode, Expression, PrimitiveTypeName, Scope, SourceFile, Statement,
    VariableSymbol, check, get_expr_type, get_field_info, get_function_info, get_let_type,
    get_struct_info,
};

fn check_source(text: &str) -> calckernel::CheckResult {
    check(&SourceFile::new("test.ck", text))
}

fn messages_of(text: &str) -> Vec<String> {
    check_source(text)
        .diagnostics
        .into_iter()
        .map(|diagnostic| diagnostic.message)
        .collect()
}

#[test]
fn check_should_accept_valid_pricing_style_program() {
    let result = check_source(
        r#"
      struct Item {
        price: i64;
        qty: i64;
        tax_rate_ppm: i64;
      }

      export fn tax(base: i64, ppm: i64) -> i64 {
        return base * ppm / 1000000;
      }

      export fn calc(items: ptr<Item>, len: i32, out: ptr<i64>) -> i32 {
        let i: i32 = 0;
        while i < len {
          let subtotal: i64 = items[i].price * items[i].qty;
          if subtotal > 0 {
            out[i] = subtotal + tax(subtotal, items[i].tax_rate_ppm);
          } else {
            out[i] = 0;
          }
          i = i + 1;
        }
        return 0;
      }
    "#,
    );

    assert_eq!(result.diagnostics, []);
    assert!(result.checked_program.function_map.contains_key("calc"));
    assert!(result.checked_program.struct_map.contains_key("Item"));
    assert!(
        result
            .typed_ast
            .expression_types
            .values()
            .any(|type_node| matches!(
                type_node,
                calckernel::CalcKernelType::Primitive(calckernel::PrimitiveTypeName::Bool)
            ))
    );
}

#[test]
fn check_should_expose_typescript_compatible_symbol_lookup_helpers() {
    let result = check_source(
        r#"
      struct Item {
        price: i64;
        qty: i32;
      }

      export fn total(items: ptr<Item>) -> i64 {
        return items[0].price;
      }
    "#,
    );

    assert_eq!(result.diagnostics, []);
    let program = &result.checked_program;
    assert_eq!(
        get_struct_info(program, "Item").expect("Item struct").name,
        "Item"
    );
    assert_eq!(
        get_field_info(program, "Item", "price")
            .expect("price field")
            .name,
        "price"
    );
    assert_eq!(
        get_function_info(program, "total")
            .expect("total function")
            .name,
        "total"
    );
    assert!(get_struct_info(program, "Missing").is_none());
    assert!(get_field_info(program, "Item", "missing").is_none());
    assert!(get_function_info(program, "missing").is_none());
}

#[test]
fn scope_should_expose_typescript_compatible_declare_and_lookup_behavior() {
    let mut scope = Scope::default();
    let outer = VariableSymbol {
        name: "value".to_string(),
        type_node: CalcKernelType::Primitive(PrimitiveTypeName::I64),
    };

    assert!(scope.declare(outer.clone()));
    assert!(!scope.declare(outer.clone()));
    assert_eq!(scope.lookup("value"), Some(&outer));
    assert_eq!(scope.lookup("missing"), None);

    let inner = VariableSymbol {
        name: "value".to_string(),
        type_node: CalcKernelType::Primitive(PrimitiveTypeName::I32),
    };
    scope.push();
    assert!(scope.declare(inner.clone()));
    assert_eq!(scope.lookup("value"), Some(&inner));
    scope.pop();
    assert_eq!(scope.lookup("value"), Some(&outer));
}

#[test]
fn check_should_report_unknown_variable_with_ck2001() {
    let result = check_source(
        r#"
      export fn bad() -> i32 {
        return missing;
      }
    "#,
    );

    assert!(result.diagnostics.iter().any(|diagnostic| {
        diagnostic.code == DiagnosticCode::Ck2001
            && diagnostic.message == "Unknown variable 'missing'."
            && diagnostic.file_name == "test.ck"
            && diagnostic.line == 3
            && diagnostic.column == 16
    }));
}

#[test]
fn check_should_report_return_type_mismatch() {
    assert!(
        messages_of(
            r#"
      export fn bad() -> i32 {
        return true;
      }
    "#
        )
        .contains(&"Return type mismatch: expected i32 but got bool.".to_string())
    );
}

#[test]
fn check_should_accept_explicit_i32_and_u32_to_f64_builtins() {
    let result = check_source(
        r#"
      export fn from_i32(n: i32) -> f64 {
        let x: f64 = i32_to_f64(n);
        return x + 1.0;
      }

      export fn from_u32(n: u32) -> f64 {
        return u32_to_f64(n);
      }
    "#,
    );

    assert_eq!(result.diagnostics, []);
}

#[test]
fn check_should_reject_strict_f64_violations() {
    assert!(
        messages_of("export fn bad() -> f64 { let x: f64 = 1; return 1.0; }")
            .contains(&"Cannot initialize 'x': expected f64 but got i32.".to_string())
    );
    assert!(
        messages_of("export fn bad(a: f64, b: i64) -> f64 { return a + b; }").contains(
            &"Arithmetic operator '+' requires integer operands of the same type.".to_string()
        )
    );
    assert!(
        messages_of("export fn bad(a: f64, b: f64) -> f64 { return a % b; }")
            .contains(&"Arithmetic operator '%' does not support f64 operands.".to_string())
    );
}

#[test]
fn check_should_expose_expression_and_let_types_for_mir_lowering() {
    let result = check_source(
        r#"
      struct Item {
        price: i64;
        qty: i64;
      }

      export fn add(a: i64, b: i64) -> i64 {
        return a + b;
      }

      export fn calc(item: ptr<Item>) -> i64 {
        let subtotal: i64 = item[0].price + add(1, 2);
        return subtotal;
      }
    "#,
    );

    assert_eq!(result.diagnostics, []);
    let calc = result
        .checked_program
        .function_map
        .get("calc")
        .expect("calc function info");
    let Statement::Let(let_statement) = &calc.declaration.body.statements[0] else {
        panic!("expected let statement");
    };
    let Expression::Binary { left, right, .. } = &let_statement.initializer else {
        panic!("expected binary initializer");
    };

    assert_eq!(
        get_let_type(&result.checked_program, let_statement),
        Some(&calckernel::CalcKernelType::Primitive(
            calckernel::PrimitiveTypeName::I64
        ))
    );
    assert_eq!(
        get_expr_type(&result.checked_program, &let_statement.initializer),
        Some(&calckernel::CalcKernelType::Primitive(
            calckernel::PrimitiveTypeName::I64
        ))
    );
    assert_eq!(
        get_expr_type(&result.checked_program, left),
        Some(&calckernel::CalcKernelType::Primitive(
            calckernel::PrimitiveTypeName::I64
        ))
    );
    assert_eq!(
        get_expr_type(&result.checked_program, right),
        Some(&calckernel::CalcKernelType::Primitive(
            calckernel::PrimitiveTypeName::I64
        ))
    );
}

#[test]
fn check_should_reject_break_and_continue_outside_while_with_ck2009() {
    let result = check_source(
        r#"
      export fn bad_break() -> i32 {
        break;
        return 0;
      }

      export fn bad_continue() -> i32 {
        continue;
        return 0;
      }
    "#,
    );
    let diagnostics = result
        .diagnostics
        .iter()
        .filter(|diagnostic| diagnostic.code.to_string() == "CK2009")
        .map(|diagnostic| {
            (
                diagnostic.message.as_str(),
                diagnostic.line,
                diagnostic.column,
            )
        })
        .collect::<Vec<_>>();

    assert_eq!(
        diagnostics,
        vec![
            ("'break' can only be used inside a while loop.", 3, 9),
            ("'continue' can only be used inside a while loop.", 8, 9),
        ]
    );
}

#[test]
fn check_should_report_unreachable_after_non_fallthrough_with_ck2010() {
    let result = check_source(
        r#"
      export fn after_return() -> i32 {
        return 0;
        return 1;
      }

      export fn after_break() -> i32 {
        while true {
          break;
          continue;
        }
        return 0;
      }

      export fn after_continue() -> i32 {
        while true {
          continue;
          break;
        }
        return 0;
      }
    "#,
    );
    let unreachable = result
        .diagnostics
        .iter()
        .filter(|diagnostic| diagnostic.code.to_string() == "CK2010")
        .map(|diagnostic| (diagnostic.line, diagnostic.column))
        .collect::<Vec<_>>();

    assert_eq!(unreachable, vec![(4, 9), (10, 11), (18, 11)]);
}

#[test]
fn check_should_combine_if_branch_flow_inside_while() {
    let result = check_source(
        r#"
      export fn branch_flow(flag: bool) -> i32 {
        while true {
          if flag {
            break;
          } else {
            continue;
          }
          let unreachable: i32 = 1;
        }

        while true {
          if flag {
            break;
          }
          let reachable: i32 = 1;
          break;
        }
        return 0;
      }
    "#,
    );
    let unreachable = result
        .diagnostics
        .iter()
        .filter(|diagnostic| diagnostic.code.to_string() == "CK2010")
        .map(|diagnostic| (diagnostic.line, diagnostic.column))
        .collect::<Vec<_>>();

    assert_eq!(unreachable, vec![(9, 11)]);
}

#[test]
fn check_should_conservatively_require_a_return_after_while_true() {
    let result = check_source(
        r#"
      export fn loop_only() -> i32 {
        while true {
          continue;
        }
      }
    "#,
    );

    assert!(
        result
            .diagnostics
            .iter()
            .any(|diagnostic| { diagnostic.message == "Missing return in function 'loop_only'." })
    );
}

#[test]
fn check_should_accept_void_fallthrough_and_empty_return() {
    let result = check_source(
        r#"
      fn no_op() -> void {}

      export fn maybe_stop(stop: bool) -> void {
        if stop {
          return;
        }
        no_op();
      }
    "#,
    );

    assert_eq!(result.diagnostics, []);
}

#[test]
fn check_should_reject_void_in_value_type_positions_with_ck2011() {
    let result = check_source(
        r#"
      struct BadField { value: void; }

      fn no_op() -> void {}
      fn bad_param(value: void) -> void {}
      fn bad_pointer(value: ptr<void>) -> void {}
      fn bad_local() -> void {
        let value: void = no_op();
      }
      fn take_i32(value: i32) -> void {}
      fn bad_argument() -> void {
        take_i32(no_op());
      }
    "#,
    );

    let void_errors = result
        .diagnostics
        .iter()
        .filter(|diagnostic| diagnostic.code.to_string() == "CK2011")
        .map(|diagnostic| diagnostic.message.as_str())
        .collect::<Vec<_>>();
    assert!(void_errors.contains(&"Void is only allowed as a function return type."));
    assert!(void_errors.contains(&"Void is not allowed as a pointer element type."));
    assert!(void_errors.contains(&"A void call cannot be used where a value is required."));
    assert!(void_errors.len() >= 5, "{void_errors:?}");
}

#[test]
fn check_should_reject_mismatched_void_returns_with_ck2011() {
    let result = check_source(
        r#"
      fn bad_void() -> void { return 1; }
      fn bad_value() -> i32 { return; }
    "#,
    );

    let errors = result
        .diagnostics
        .iter()
        .filter(|diagnostic| diagnostic.code.to_string() == "CK2011")
        .map(|diagnostic| diagnostic.message.as_str())
        .collect::<Vec<_>>();
    assert_eq!(
        errors,
        vec![
            "A void function cannot return a value.",
            "A non-void function must return a value.",
        ]
    );
}

#[test]
fn check_should_accept_only_void_calls_as_statements() {
    let accepted = check_source(
        r#"
      fn no_op() -> void {}
      export fn run() -> void { no_op(); }
    "#,
    );
    assert_eq!(accepted.diagnostics, []);

    let rejected = check_source(
        r#"
      fn value() -> i32 { return 1; }
      export fn run() -> void { value(); }
    "#,
    );
    assert!(rejected.diagnostics.iter().any(|diagnostic| {
        diagnostic.code.to_string() == "CK2011"
            && diagnostic.message == "Only a void call may be used as a standalone statement."
    }));
}

#[test]
fn check_should_reject_void_call_in_value_contexts() {
    let result = check_source(
        r#"
      fn no_op() -> void {}
      fn take(value: i32) -> i32 { return value; }
      fn initializer() -> i32 { let x: i32 = no_op(); return 0; }
      fn returned() -> i32 { return no_op(); }
      fn argument() -> i32 { return take(no_op()); }
      fn arithmetic() -> i32 { return no_op() + 1; }
      fn comparison() -> bool { return no_op() == no_op(); }
    "#,
    );

    let void_value_errors = result
        .diagnostics
        .iter()
        .filter(|diagnostic| {
            diagnostic.code.to_string() == "CK2011"
                && diagnostic.message == "A void call cannot be used where a value is required."
        })
        .count();
    assert_eq!(void_value_errors, 6);
}

#[test]
fn check_should_preserve_unreachable_analysis_after_empty_return() {
    let result = check_source(
        r#"
      export fn stop() -> void {
        return;
        no_op();
      }
      fn no_op() -> void {}
    "#,
    );

    assert!(result.diagnostics.iter().any(|diagnostic| {
        diagnostic.code.to_string() == "CK2010"
            && diagnostic.message == "Unreachable statement."
            && diagnostic.line == 4
    }));
}
